src/sparsezoo/v2/directory.py

```diff
@@ -0,0 +1,222 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import pathlib
+import tarfile
+from typing import List, Optional
+
+from sparsezoo.utils.downloader import download_file
+from sparsezoo.v2.file import File, retry
+
+
+__all__ = ["Directory"]
+
+
+class Directory(File):
+    """
+    Object that represents a directory.
+    Directory may also represent a tar file.
+
+    :param files: list of files contained within the Directory
+        (None for tar archive files)
+    :param name: name of the Directory
+    :param path: path of the Directory
+    :param url: url of the Directory
+    """
+
+    def __init__(
+        self,
+        name: str,
+        files: Optional[List[File]] = None,
+        path: Optional[str] = None,
+        url: Optional[str] = None,
+    ):
+
+        self.files = files
+        extension = name.split(".")[-2:]
+        self._is_archive = (extension == ["tar", "gz"]) and (not self.files)
+
+        super().__init__(name=name, path=path, url=url)
+
+        if self._unpack():
+            self.unzip()
+
+    @property
+    def is_archive(self) -> bool:
+        """
+        Boolean flag:
+        - True if the Directory is an archive (tar)
+        - False if the Directory is a local directory (folder)
+
+        :return: boolean flag; True if Directory is tar archive, False if folder
+        """
+        return self._is_archive
+
+    @is_archive.setter
+    def is_archive(self, value: bool):
+        """
+        Setter property of `is_archive`
+
+        :param value: boolean flag; True if Directory is tar archive, False if folder
+        """
+        self._is_archive = value
+
+    def get_file_names(self) -> List[str]:
+        """
+        Get the names of the files in the Directory
+
+        return: List with names of files
+        """
+        if self.is_archive:
+            tar = tarfile.open(self.path)
+            return [os.path.basename(member.name) for member in tar.getmembers()]
+        else:
+            return [file.name for file in self.files]
+
+    def validate(self, strict_mode: bool = True):
+        """
+        Validates the structure of directory.
+
+        :return: boolean flag; True if files are valid and no errors arise
+        """
+
+        if self.is_archive:
+            # we are not validating tar files for now
+            validations = {self.name: True}
+        else:
+            validations = {}
+            for file in self.files:
+                validations[file.name] = file.validate(strict_mode=strict_mode)
+
+        if not all(validations.values()):
+            logging.warning(
+                "Following files: "
+                f"{[key for key, value in validations.items() if not value]} "
+                "were not successfully validated."
+            )
+
+        return all(validations.values())
+
+    @retry(retries=3, retry_sleep_sec=5)
```

Comment:
@rahul-tuli @dbogunowicz great idea here - my worry is making this default behavior that the user can't override.  Let's get this added in instead as a parameter to download (can use recursion or something similar), maybe default to 1 that way the user can control the amount of retries and not take a max of 15 seconds + time to fail before exiting by default

Reply:
you can probably keep the retry decorator, just have it read max retries in from the function parameters and default to 1